### ISSUE TITLE
"hello world" example with the `wasi` crate doesn't work

### ISSUE THREAD

**Issue by yoshuawuyts** [bytecodealliance/cargo-component#236]
## Repro

If we run the following commands to build the [wasi crate](https://github.com/bytecodealliance/wasi/tree/main):
```bash
cargo install cargo-component
git clone https://github.com/bytecodealliance/wasi
cd wasi
cargo component build --example hello-world
```

This results in the following output:

```text
yosh@MacBook-Pro wasi % cargo component run --example hello-world
   Compiling wit-bindgen v0.19.1
   Compiling bitflags v2.4.2
   Compiling wasi v0.12.1+wasi-0.2.0 (/Users/yosh/Code/wasi)
    Finished dev [unoptimized + debuginfo] target(s) in 1.76s
Error: failed to run main module `/Users/yosh/Code/wasi/target/wasm32-wasi/debug/examples/hello-world.wasm`

Caused by:
    0: failed to instantiate "/Users/yosh/Code/wasi/target/wasm32-wasi/debug/examples/hello-world.wasm"
    1: unknown import: `wasi:io/streams@0.2.0::[resource-drop]output-stream` has not been defined
```

## Manual build instructions

The `wasi` crate suggests the following instructions to get it to build correctly:

```bash
# build the example for WASI 0.1
cargo build --target wasm32-wasi --example hello-world

# fetch the WASI 0.1 -> WASI 0.2 adapter
curl -LO https://github.com/bytecodealliance/wasmtime/releases/download/v17.0.0/wasi_snapshot_preview1.command.wasm

# convert the example into a WASI 0.2 component
cargo install wasm-tools
wasm-tools component new target/wasm32-wasi/example/hello-world.wasm \
    --adapt ./wasi_snapshot_preview1.command.wasm \
    -o hello-world.wasm

# run the component
wasmtime run hello-world.wasm
```

This results in the following output:

```text
Hello, world!
```

## Expected behavior

From the outset I was expecting this to just work? Is it that `cargo run --example` isn't implemented yet? It does seem to build the `wasm32-wasi` example correctly. Or is something else perhaps not working when it should?

**Comment by peterhuene** [comment]
To prevent `cargo-component` from componentizing any ol' `wasm32` target it encounters, it expects `Cargo.toml` to have the `[package.metadata.component]` section to treat it as a component rather than a module.

If you add `[package.metadata.component]` to `Cargo.toml`, you'll get errors when parsing the WIT, as `cargo-component` does not do discovery via a `deps` directory, instead expects entries in `[package.metadata.component.target.dependencies]`.

So `Cargo.toml` would need:

```toml
[package.metadata.component.target.dependencies]
"wasi:http" = { path = "wit/deps/http" }
"wasi:random" = { path = "wit/deps/random" }
"wasi:cli" = { path = "wit/deps/cli" }
"wasi:io" = { path = "wit/deps/io" }
"wasi:clocks" = { path = "wit/deps/clocks" }
"wasi:filesystem" = { path = "wit/deps/filesystem" }
"wasi:sockets" = { path = "wit/deps/sockets" }
```

However, there appears to be breaking changes in the latest `wit-bindgen` with regards to runtime types, so you'd need to drop the version down to `0.18.0` to be compatible with the latest `cargo-component` (I should do a release to bump it).

With the above in place, you should now get:

```
Error: failed to run main module `/Users/peterhuene/tmp/wasi/target/wasm32-wasi/debug/examples/hello-world.wasm`

Caused by:
    exported instance `wasi:cli/run@0.2.0` not present
```

This, finally, is a bug in `cargo-component`, where it is not treating an example as a bin, and thus is using the built-in reactor adapter rather than the command adapter.

It's possible to workaround that with:

```toml
[package.metadata.component]
adapter = "wasi_snapshot_preview1.command.wasm"
```

But it should work out of the box.

**Comment by alexcrichton** [comment]
> it expects Cargo.toml to have the [package.metadata.component] section to treat it as a component rather than a module

In this case with no WIT internally as it's all coming from the `wasi` crate itself could this logic change to: if no `[package.metadata.component]` is specified then parse the output wasm module and if any `component-type*` custom section is found then do the componentization process as-if an empty section was specified?

> However, there appears to be breaking changes in the latest wit-bindgen with regards to runtime types

Oh dear, do you mean breakage from 0.18-to-0.19 or something that needs a yank/republish on crates.io?

(also I'm happy to cc you on more changes to wit-bindgen if you'd like for possible breakage with cargo-component)

**Comment by alexcrichton** [comment]
Ah I see good point!

One suggestion would be to update `cargo component` to generate `wit-bindgen = "0.18.0"` in the manifest, or whatever matches the built-in generator of `cargo component`. That being said that won't play well with upgrades over time, for example taking a very new `cargo component` and building an older project. 

One fix for that is to then do what you mention, generate `Reource<T>` and bits into the file itself. That has problems though we just discovered at least with `cabi_realloc` where we want multiple wit-bindgen-using crates to link together and they can't all define `cabi_realloc` so we need to use weak symbols, and then that's not stable in Rust.

I think the best solution might be to define "perma stable" APIs, like how the runtime links `cabi_realloc` and the `run_ctors_once` thing. Less stable things, like `Resource<T>` should go into the generated bindings.

So how about this:

* As a near-term fix, `cargo component` updates to 0.19.x for `wit-bindgen` generation
* As a long-term fix, `wit-bindgen` generates the `Resource<T>` wrapper and any "less stable" bits into the generated bindings
* As a long-term fix, `wit-bindgen` considers its API surface are "perma stable" as future versions of `wit-bindgen` runtime should work with old versions of `wit-bindgen` generators

**Comment by peterhuene** [comment]
Your plan sounds great to me!
